### PR TITLE
Fix Map.getOption nested null bug

### PR DESCRIPTION
**Pull request by nomisRev**
This PR fixes a bug with _nested null_, or when `V` is `null`, in `Map.getOption`.

### CODE REVIEW

File: lib/src/test/kotlin/app/cash/quiver/extensions/MapTest.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import arrow.core.Some
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.shouldBe
+
+class MapTest : StringSpec({
+  "Can retrieve null value key" {
+    mapOf(
+      "one" to 1,
+      "null" to null
+    ).getOption("null") shouldBe Some(null)
```
Comment (cwmyers) [inline lib/src/test/kotlin/app/cash/quiver/extensions/MapTest.kt:12]: I don't think this test captures the intent of this function.  We intentionally wanted any null values be returned as `None`.  i.e. we don't distinguish between value absence and a value being `null`.

Otherwise the signature should be  `fun <K, A> Map<K, A>.getOption(k: K): Option<A?>` which we wouldn't find as useful.

Comment (cwmyers) [reply lib/src/test/kotlin/app/cash/quiver/extensions/MapTest.kt:12]: Can you expound a bit more on your thinking here @nomisRev? I'm conscious that making assumptions about underlying values is dangerous and agree your implementation preserves the reality of the data inside the map.

Comment (nomisRev) [reply lib/src/test/kotlin/app/cash/quiver/extensions/MapTest.kt:12]: Hey @cwmyers,

Yes, of course. For reference, it's similar to the fix I created last week for [`filterNotNone`](https://github.com/cashapp/quiver/pull/22/files).

> i.e. we don't distinguish between value absence and a value being null.

That can of course be a valid use-case but it's not the typical behavior of this function. It's perhaps more obvious in Kotlin due to having nullable types.

> Otherwise the signature should be  fun <K, A> Map<K, A>.getOption(k: K): Option<A?> which we wouldn't find as useful.

This doesn't seem correct IMO, because when `A : Any` is `Int` for example then you would return `Int?` when that is in fact never possible. So the problem only arises when you use `Int?` as a type for `A` in `Map<K, A>`.

If we take a look at the documentation of [`Map.get`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-map/get.html) it says: "Returns the value corresponding to the given [key](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-map/get.html#kotlin.collections.Map$get(kotlin.collections.Map.K)/key), or null if such a key is not present in the map.".

If we translate that to `Option`: "Returns the value corresponding to the given [key](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-map/get.html#kotlin.collections.Map$get(kotlin.collections.Map.K)/key) as `Some`, or `None` if such a key is not present in the map.".

Which is what I modelled with my fix. The reasoning is that if you work with `Map` in generic code, you never know if the users code is non-null (`Int`) or nullable (`Int?`). Using `Option` saves you from this pitfall in generic code, but only with the suggested implementation. Otherwise you could _falsely assume the key doesn't exist_, which results in subtle bugs. [Scala has the same behavior](https://scastie.scala-lang.org/LP4YcAfLRZu0hvYg8QwEHQ).

If this is not the desired behavior, perhaps an alternative implementation could be to constraint `A` to `Any` using `A : Any`. That avoids this problem all together, but restricts the usage of the function to only non-null types.

TL;DR: `getOption` typically returns `Some` if a `K` is found, and `None` if no `K` is found. Not doing so can lead to false assumption that no user value is present in generic code when `A` is not constraint to `A : Any`.

Comment (cwmyers) [reply lib/src/test/kotlin/app/cash/quiver/extensions/MapTest.kt:12]: Yep I get what you are saying and I agree.  I forgot that the `A` could represent something like a `Int?` which means we would be making false judgements about the data.

I'll improve the documentation around this to make this clearer.  I wonder also if any of our code internally is explicitly relying on this behaviour.

**Review by cwmyers — APPROVED** [review]
